src/storage/ring_buffer.rs

```diff
@@ -101,6 +119,67 @@ impl<'a, T: 'a> RingBuffer<'a, T> {
             Err(error) => Err(error)
         }
     }
+
+    /// Get capacity of the underlying storage.
+    pub fn capacity(&self) -> usize {
+        self.storage.len()
+    }
+
+    /// Remove the first element equal to `value` from the buffer,
+    /// reutrn `Err(Error::Exhausted)` if no such element was found.
+    pub fn remove(&mut self, value: &T) -> Result<()>
+        where T: Eq + Copy {
+        if self.empty() {
+            return Err(Error::Exhausted);
+        }
+
+        let mut p = None;
+        for i in 0..self.length {
+            if self.storage[self.mask(self.read_at + i)] == *value {
+                p = Some(i);
+                break;
+            }
+        }
+
+        if let Some(i) = p {
+            for j in i..self.length-1 {
+                self.storage[self.mask(self.read_at + j)] =
+                    self.storage[self.mask(self.read_at + j + 1)];
+            }
+            self.length -= 1;
+            Ok(())
+        } else {
+            Err(Error::Exhausted)
+        }
+    }
+
+    /// Try to expand the underlying storage, panic if the storage cannot be expanded.
+    #[cfg(not(any(feature = "std", feature = "collections")))]
+    pub fn expand_storage(&mut self) {
+        panic!("Cannot expand a ring_buffer in not-owned memory")
+    }
+
+    #[cfg(any(feature = "std", feature = "collections"))]
+    pub fn expand_storage(&mut self)
+        where T: Copy + Default {
```

Comment:
Do you need `Copy`? I think `mem::swap` or even `slice::swap` is enough to implement the expansion.

Reply:
I guess that's twice as much work so the Copy bound is fine.